Fix selinux denials for host path e2e tests.

aveshagarwal opened the pull request:
This PR fixes selinux denials as the containers are accessing a subdir in /tmp as hostpath volume source for which selinux relabeling is not supported for these volumes so making them privileged.

```
time->Thu Aug 25 20:18:26 2016
type=AVC msg=audit(1472141906.557:3523): avc:  denied  { write } for  pid=15682 comm="mt" name="test-file" dev="tmpfs" ino=317712 scontext=system_u:system_r:svirt_lxc_net_t:s0:c29,c183 tcontext=system_u:object_r:docker_tmp_t:s0 tclass=file permissive=0
----
time->Thu Aug 25 20:18:27 2016
type=AVC msg=audit(1472141907.814:3527): avc:  denied  { open } for  pid=15771 comm="mt" path="/test-volume/sub-path/test-file" dev="tmpfs" ino=317712 scontext=system_u:system_r:svirt_lxc_net_t:s0:c143,c496 tcontext=system_u:object_r:docker_tmp_t:s0 tclass=file permissive=0
----
time->Thu Aug 25 20:18:29 2016
type=AVC msg=audit(1472141909.814:3528): avc:  denied  { open } for  pid=15771 comm="mt" path="/test-volume/sub-path/test-file" dev="tmpfs" ino=317712 scontext=system_u:system_r:svirt_lxc_net_t:s0:c143,c496 tcontext=system_u:object_r:docker_tmp_t:s0 tclass=file permissive=0
```

@kubernetes/rh-cluster-infra @pmorie

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.kubernetes.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/31542)

<!-- Reviewable:end -->

ncdc commented on test/e2e/common/host_path.go:
```diff
@@ -133,6 +133,7 @@ func mount(source *api.HostPathVolumeSource) []api.Volume {
 func testPodWithHostVol(path string, source *api.HostPathVolumeSource) *api.Pod {
 	podName := "pod-host-path-test"
 
+	priv := true
```
I really don't want to run the container privileged if we can avoid it. Is there any other way to do this? Can we create a temporary directory, chcon it to svirt_sandbox_file_t, and then run these tests?

aveshagarwal replied:
I just noticed that no test is doing that. See test/e2e/volumes.go too.

ncdc replied:
I would consider those volumes tests in a different class (they need
privileged to do thinks like mounting). You shouldn't need to be privileged
just to get a hostMount to work.

On Fri, Aug 26, 2016 at 3:20 PM, Avesh Agarwal <notifications@github.com>
wrote:

> In test/e2e/common/host_path.go
> <https://github.com/kubernetes/kubernetes/pull/31542#discussion_r76473301>
> :
>
> > @@ -133,6 +133,7 @@ func mount(source *api.HostPathVolumeSource) []api.Volume {
> >  func testPodWithHostVol(path string, source *api.HostPathVolumeSource) *api.Pod {
> >  	podName := "pod-host-path-test"
> >
> > +	priv := true
>
> I just noticed that no test is doing that. See test/e2e/volumes.go too.
>
> —
> You are receiving this because you are on a team that was mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/31542/files/d5a5fd68e797f42ab582ac9964750b0277b27a22#r76473301>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAABYjJ857PXzBoIn12cs-R_3N6gJ6DTks5qjzyEgaJpZM4JuYK8>
> .
>

fejta commented:
@ncdc assigning to you based on comments! Please let me know if you need something from me.

ncdc commented:
@pmorie how do you want to handle this?

aveshagarwal commented:
I am testing code changes by relabeling only the specific dir instead of using priv mode. Will update soon.

aveshagarwal commented:
@pmorie @ncdc updated. PTAL.

ncdc commented on test/e2e/common/host_path.go:
```diff
@@ -169,3 +181,27 @@ func testPodWithHostVol(path string, source *api.HostPathVolumeSource) *api.Pod
+func relabelPathByChcon(path string) error {
+	dir, err := filepath.Abs(path)
+	if err != nil {
+		return fmt.Errorf("error converting directory '%s' to an absolute path: %v", path, err)
```
%q prints out a quoted string

ncdc commented on test/e2e/common/host_path.go:
```diff
@@ -169,3 +181,27 @@ func testPodWithHostVol(path string, source *api.HostPathVolumeSource) *api.Pod
+
+	if chconPath, err := exec.LookPath("chcon"); err != nil {
+		return fmt.Errorf("couldn't locate the command chcon on host: %v", err)
+	} else {
```
Don't need this inside an else block

ncdc commented on test/e2e/common/host_path.go:
```diff
@@ -86,12 +90,20 @@ var _ = framework.KubeDescribe("HostPath", func() {
+		if selinux.SelinuxEnabled() {
+			filePath := path.Join(hostPath, subPath)
+			if err := relabelPathByChcon(filePath); err != nil {
+				glog.Warningf("%v", err)
```
Would `t.Logf` be better here? You also might want to include some context ("Error relabeling %q: %v", filePath, err)

aveshagarwal replied:
As relabelPathByChcon returns error context from all possible errors paths, did not include context here to avoid duplication.

ncdc commented on test/e2e/common/host_path.go:
```diff
@@ -169,3 +181,28 @@ func testPodWithHostVol(path string, source *api.HostPathVolumeSource) *api.Pod
+
+	if _, err := os.Stat(dir); os.IsNotExist(err) {
+		if err := os.MkdirAll(dir, 0755); err != nil {
+			return fmt.Errorf("couldn't create directory on host '%s': %s", dir, err)
```
%q instead of '%s'

Commit:
Fix host path e2e on selinux enabled systems by relabeling selinux context of
a sub dir in the /tmp dir to svirt_sandbox_file_t. It fixes specifically
following e2e test: "should support subPath [Conformance]"

Commit:
Fix host path e2e on selinux enabled systems by relabeling selinux context of
a file in the /tmp dir to svirt_sandbox_file_t. It fixes specifically following
e2e test: "should support r/w".

aveshagarwal commented:
@ncdc I have reworked this PR to relabel to not only dir but also files. By doing this, also handles another test case failure "should support r/w". So this pr now addresses 2 test case failures:

`"should support subPath [Conformance]"`
`should support r/w`

aveshagarwal commented:
@k8s-bot test this issue #31083

stevekuznetsov commented:
@pmorie this looks like the fix we need -- can we try to get it in?
